app/main.py

```diff
@@ -1,213 +1,37 @@
+def prepare_knight(knight: dict) -> None:
+    knight["protection"] = sum(a["protection"] for a in knight["armour"])
+    knight["power"] += knight["weapon"]["power"]
+    if knight["potion"]:
+        for effect in ["power", "protection", "hp"]:
+            if effect in knight["potion"]["effect"]:
+                knight[effect] += knight["potion"]["effect"][effect]
 
-    # mordred
-    mordred = knightsConfig["mordred"]
 
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
+def battle(knights_config: dict) -> dict:
+    for knight_name, knight in knights_config.items():
+        prepare_knight(knight)
 
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+    lancelot = knights_config["lancelot"]
+    mordred = knights_config["mordred"]
+    arthur = knights_config["arthur"]
+    red_knight = knights_config["red_knight"]
 
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
     lancelot["hp"] -= mordred["power"] - lancelot["protection"]
     mordred["hp"] -= lancelot["power"] - mordred["protection"]
 
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
     arthur["hp"] -= red_knight["power"] - arthur["protection"]
     red_knight["hp"] -= arthur["power"] - red_knight["protection"]
 
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
+    for knight in [lancelot, mordred, arthur, red_knight]:
+        knight["hp"] = max(0, knight["hp"])
```

Comment:
Place this logic in a separate function and place it in another file, so that yo will pass two knights instances and will receive their hp as a result of a fight.